File: ips/ips-model/src/main/java/io/iochord/apps/ips/data/services/DataConnectionService.java
package io.iochord.apps.ips.data.services;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.concurrent.CompletableFuture;

import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import com.opencsv.CSVParserBuilder;
import com.opencsv.CSVReader;
import com.opencsv.CSVReaderBuilder;

import io.iochord.apps.ips.data.services.models.ImportCsvConfiguration;
import io.iochord.apps.ips.data.services.models.ImportCsvResult;
import io.iochord.apps.ips.services.AService;
import io.iochord.apps.ips.services.ServiceExecutor;
import io.iochord.apps.ips.services.ServiceState;
import io.iochord.apps.ips.util.SerializationUtil;

/**
 *
 * @package chdsr-simulator-web
 * @author  Iq Reviessay Pulshashi <pulshashi@ideas.web.id>
 * @since   2019
 *
 *
 */
@Service
public class DataConnectionService extends AService {
	
	@Async(ServiceExecutor.NAME)
	public CompletableFuture<ImportCsvResult> importCsv(ImportCsvConfiguration config, ServiceState state) {
		String name = "dataset_" + state.getSessionId();
		ImportCsvResult result = new ImportCsvResult();
		result.setConfig(config);
		result.setName(name);
		try {
			Connection conn = getDataSource().getConnection();
			CSVReader csvReader = new CSVReaderBuilder(config.getReader())
				.withCSVParser(
					new CSVParserBuilder()
					.withSeparator(config.getDelimeter())
					.build())
				.build();
			for (int i = 0; i < config.getHeaderRowIdx(); i++) {
				csvReader.readNext();
			}
			String[] cells = csvReader.readNext();
			int rows = 0;
			if (cells != null) {
				int cols = cells.length;
				StringBuilder sql = new StringBuilder();
				StringBuilder isql = new StringBuilder();
				sql.append("CREATE TABLE ").append(name).append(" ( eid SERIAL PRIMARY KEY ");
				isql.append("INSERT INTO ").append(name).append(" VALUES ( DEFAULT ");
				for (int i = 0; i < cols; i++) {
					sql.append(",");
					isql.append(",");
					sql.append("c").append(i).append(" VARCHAR(255) NULL");
					isql.append("?");
				}
				sql.append(");");
				isql.append(");");
				try {
					PreparedStatement st = conn.prepareStatement(sql.toString());
					st.execute();
					st.close();
					sql = new StringBuilder();
					sql.append("COMMENT ON TABLE ").append(name).append(" IS '").append(SerializationUtil.encode(config)).append("';");
					st = conn.prepareStatement(sql.toString());
					st.execute();
					st.close();
				} catch (SQLException e) {
					e.printStackTrace();
				}
				PreparedStatement st = conn.prepareStatement(isql.toString());
				int brows = 0;
				while (cells != null) {
					for (int i = 0; i < cols; i++) {
						st.setString(i + 1, i < cells.length ? cells[i] : null);
					}
					st.addBatch();
					if (brows % 1000 == 0) {
						getWsmTemplate().convertAndSend(state.getProgressWsUri(), rows + " / ");
						st.executeBatch();
						st.close();
						st = conn.prepareStatement(isql.toString());
						brows = 0;
					}
					cells = csvReader.readNext();
					rows++;
					brows++;
				}
				if (brows > 0) {
					st.executeBatch();
				}
				st.close();
			}
			csvReader.close();
			conn.close();
			state.setStatus(ServiceState.STATE.COMPLETED);
			getWsmTemplate().convertAndSend(state.getCompleteWsUri(), result);
		} catch (Exception e) {
			e.printStackTrace();
		}
		return CompletableFuture.completedFuture(result);
	}

}

File: ips/ips-simulator-web/src/main/ts/src/iochord/ips/data/connection/services/DataConnectionService.ts
import DataService from '../../services/DataService';
import ImportCsvConfiguration from '../models/ImportCsvConfiguration';

export { ImportCsvConfiguration };

/**
 * The data connection service class.
 *
 * @export
 * @class DataConnectionService
 * @extends {DataService}
 *
 * @package ts
 * @author Iq Reviessay Pulshashi <pulshashi@ideas.web.id>
 * @since 2019
 */
export default class DataConnectionService extends DataService {

  /**
   * Base URI for data connection service.
   *
   * @static
   * @type {string}
   * @memberof DataConnectionService
   */
  public static readonly BASE_URI: string = DataService.BASE_URI + '';

  /**
   * Returns the instance of data connection service.
   *
   * @static
   * @returns {DataConnectionService}
   * @memberof DataConnectionService
   */
  public static getInstance(): DataConnectionService {
    if (DataConnectionService.__INSTANCE == null) {
      DataConnectionService.__INSTANCE = new DataConnectionService();
    }
    return DataConnectionService.__INSTANCE;
  }

  /**
   * The instance of data connection service.
   *
   * @private
   * @static
   * @type {DataConnectionService}
   * @memberof DataConnectionService
   */
  private static __INSTANCE: DataConnectionService;

  /**
   * Returns the data connection list.
   *
   * @param {*} completeCallback
   * @param {*} progressCallback
   * @returns
   * @memberof DataConnectionService
   */
  public getDataConnections(completeCallback: any, progressCallback: any) {
    return this.webserviceGet(DataConnectionService.BASE_URI + '/connection/list', completeCallback, progressCallback);
  }

  public getFilteredDataConnections(type: any, completeCallback: any, progressCallback: any) {
    return this.webserviceGet(DataConnectionService.BASE_URI + '/connection/list/' + type, completeCallback, progressCallback);
  }

  /**
   * Import CSV to the web service.
   *
   * @param {ImportCsvConfiguration} config
   * @param {*} file
   * @param {*} completeCallback
   * @param {*} progressCallback
   * @returns
   * @memberof DataConnectionService
   */
  public importCsv(config: ImportCsvConfiguration, file: any, completeCallback: any, progressCallback: any) {
    const req = new FormData();
    req.append('config', JSON.stringify(config));
    req.append('file', file);
    return this.webserviceUpload(DataConnectionService.BASE_URI + '/import/csv', req, completeCallback, progressCallback);
  }

}
